### PR TITLE
chore: create packages

### PR DESCRIPTION
**Pull request by crazy-max**
Follow-up to our discussion about releasing buildx as a separate package.

I first started to implement this on the [docker-ce-packaging](https://github.com/docker/docker-ce-packaging) repo but the [Go version](https://github.com/docker/docker-ce-packaging/blob/8bbb124d84a94765592fcb4aeef66982daf2d69d/common.mk#L19) is not the one we want for buildx. I think we should first merge docker/cli#3258.

Creating packages here allows for easier maintenance. In a follow-up we could upload pkgs and update repodata on `download.docker.com` (cc @StefanScherer).

Test it:

```console
$ docker buildx bake "https://github.com/crazy-max/buildx.git#deb-pkg" pkg
```

```console
$ dpkg-deb -I bin/docker-buildx_0.7.0~9-g1250d945.m_amd64.deb 
 new Debian package, version 2.0.
 size 12219872 bytes: control archive=419 bytes.
       1 bytes,     0 lines      conffiles
     306 bytes,    10 lines      control
      81 bytes,     1 lines      md5sums
 Package: docker-buildx
 Version: 0:0.7.0~9-g1250d945.m
 Section: default
 Priority: optional
 Architecture: amd64
 Maintainer: Docker <support@docker.com>
 Vendor: Docker
 Installed-Size: 46008
 Homepage: https://github.com/docker/buildx
 Description: Docker Buildx plugin extends build capabilities with BuildKit.
```

```console
$ rpm -qip bin/docker-buildx-0.7.0~9-g1250d945.m.x86_64.rpm 
Name        : docker-buildx
Epoch       : 0
Version     : 0.7.0~9-g1250d945.m
Release     : 1
Architecture: x86_64
Install Date: (not installed)
Group       : Tools/Docker
Size        : 47112192
License     : Apache-2.0
Signature   : (none)
Source RPM  : docker-buildx-0.7.0~9-g1250d945.m-1.src.rpm
Build Date  : Thu Dec 16 11:31:29 2021
Build Host  : buildkitsandbox
Relocations : (not relocatable)
Packager    : Docker <support@docker.com>
Vendor      : Docker
URL         : https://github.com/docker/buildx
Summary     : Docker Buildx plugin extends build capabilities with BuildKit.
Description :
Docker Buildx plugin extends build capabilities with BuildKit.
```

cc @thaJeztah @tonistiigi

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -59,12 +59,59 @@ FROM --platform=$BUILDPLATFORM alpine AS releaser
...
 FROM scratch AS release
 COPY --from=releaser /out/ /
 
+# Deb
+FROM --platform=$BUILDPLATFORM debian:bullseye-slim AS build-deb
+RUN apt-get update && apt-get install -y debhelper dh-make
+WORKDIR /build
+COPY --from=xx / /
+COPY --from=build /usr/bin/buildx ./docker-buildx-standalone/usr/bin/
+COPY --from=build /usr/bin/buildx ./docker-buildx-plugin/usr/libexec/docker/cli-plugins/docker-buildx
```
Comment (tonistiigi) [inline Dockerfile:74]: I don't think we need two packages. We can just put the binary at `/usr/libexec/docker/cli-plugins/docker-buildx` and symlink it to `/usr/bin/buildx`.

File: Dockerfile [hunk 2]
```diff
@@ -59,12 +59,59 @@ FROM --platform=$BUILDPLATFORM alpine AS releaser
...
+ARG TARGETPLATFORM
+RUN --mount=type=bind,source=/tmp/.version,target=/tmp/.version,from=version <<EOT
+#!/usr/bin/env bash
+set -e
+version=$(cat /tmp/.version)
+if [[ $version =~ ^[a-f0-9]{7}$ ]]; then
+  version="v0.0.0+${version}"
+fi
+mkdir -p ./docker-buildx-standalone/DEBIAN
+cat > ./docker-buildx-standalone/DEBIAN/control <<EOL
```
Comment (tonistiigi) [inline Dockerfile:84]: That seems a bit too much data even for heredocs. Maybe create a template file under contrib or dist.

Comment (crazy-max) [reply Dockerfile]: Switched to nfpm so we can handle deb and rpm packages with the same config. Maybe apk is not necessary as buildx is already available for alpine: https://pkgs.alpinelinux.org/packages?name=*buildx*&branch=v3.14

File: docker-bake.hcl [hunk 3]
```diff
@@ -122,12 +122,38 @@ target "binaries-cross" {
   ]
 }
 
-target "release" {
+target "pkg" {
+  inherits = ["binaries"]
+  args = {
+    PKG = "apk,deb,rpm"
```
Comment (crazy-max) [inline docker-bake.hcl:128]: Maybe apk is not necessary as buildx is already available for alpine: https://pkgs.alpinelinux.org/packages?name=*buildx*&branch=v3.14

File: Dockerfile [hunk 4]
```diff
@@ -59,12 +60,44 @@ FROM --platform=$BUILDPLATFORM alpine AS releaser
...
+set -e
+version=$(cat /tmp/.version)
+if [[ $version =~ ^[a-f0-9]{7}$ ]]; then
+  version="v0.0.0+${version}"
+fi
+for pkg in ${PKG//,/ }
+do
+  arch=$(xx-info march)
+  [[ "$pkg" = "apk" ]] && arch=$(xx-info alpine-arch)
+  [[ "$pkg" = "deb" ]] && arch=$(xx-info debian-arch)
```
Comment (crazy-max) [inline Dockerfile:91]: needs https://github.com/tonistiigi/xx/pull/52 to handle rhel archs properly.

File: docker-bake.hcl (hunk 3 above)
Comment (tomalok) [reply docker-bake.hcl:128]: fwiw, here's the `APKBUILD` that we use.  we regularly rebuild for new `go` versions, CVEs, etc.

https://gitlab.alpinelinux.org/alpine/aports/-/blob/master/community/docker-cli-buildx/APKBUILD

Comment (crazy-max) [reply docker-bake.hcl]: @tomalok sgtm thanks for your feedback

File: docker-bake.hcl [hunk 5]
```diff
@@ -122,12 +122,38 @@ target "binaries-cross" {
   ]
 }
 
-target "release" {
+target "pkg" {
+  inherits = ["binaries"]
+  args = {
+    PKG = "deb,rpm"
```
Comment (crazy-max) [inline docker-bake.hcl:128]: @tonistiigi As discussed only create deb and rpm pkgs

File: hack/nfpm.yml [hunk 6]
```diff
@@ -0,0 +1,30 @@
+name: docker-buildx
```
Comment (crazy-max) [inline hack/nfpm.yml:1]: Do we have a convention for package names? I see docker scan is [named `docker-scan-plugin`](https://github.com/docker/docker-ce-packaging/blob/8bbb124d84a94765592fcb4aeef66982daf2d69d/deb/common/control#L105). Maybe we should do the same here with `docker-buildx-plugin`? cc @thaJeztah

Comment (tonistiigi) [reply hack/nfpm.yml:1]: Based on https://github.com/docker/cli/pull/3314#issuecomment-945667368 we are distancing from "plugin" lingo so would rather avoid it. The bundled tools by Docker org that make up "docker" are better referred as components. Compose is compose, not compose-plugin. Same for buildx.

File: hack/nfpm.yml [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+platform: linux
+
+version: ${VERSION}
+epoch: 0
+
+section: default
+
+maintainer: Docker <support@docker.com>
+description: Docker Buildx plugin extends build capabilities with BuildKit.
+vendor: Docker, Inc
```
Comment (AkihiroSuda) [inline hack/nfpm.yml:16]: Maybe the vendor name should be provided via `--build-arg`?

Comment (crazy-max) [reply hack/nfpm.yml:16]: Hum yeah indeed, thanks for your feedback

Comment (crazy-max) [reply hack/nfpm.yml]: needs https://github.com/goreleaser/nfpm/pull/406 to expand env for vendor field or we could sed but that's kinda dirty.

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -46,6 +46,10 @@ jobs:
         name: Build binaries
         run: |
           make release
+      -
+        name: Build packages
+        run: |
+          make pkg
```
Comment (crazy-max) [inline .github/workflows/build.yml:52]: Packages will not be included in GitHub Releases (only build). Instead we want to use the remote bake definition from docker-ce-packaging repo.

**Comment by thaJeztah** [comment]
So IIUC, the idea is for the buildx repository to provide the "recipe" for building the `.deb` and `.rpm` packages, and to update the `docker-ce-packaging` pipeline, to run it as part of releases, correct?

I had a look at what it produces, and noticed that this currently produces static binaries (packaged as `.deb` / `.rpm`).

```bash
ldd /usr/libexec/docker/cli-plugins/docker-buildx
	linux-vdso.so.1 (0x00007fffc1dc7000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fa7e9f0a000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fa7e9b19000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fa7ea129000)
```

I know the current buildx binaries we ship in the `docker-ce-cli` packages are dynamically linked, and perhaps that's not an issue, but I do recall that @justincormack didn't want to be using static binaries (and I also recall some binaries could not be shipped as static binaries on Docker Desktop, related to some GPL code, but not sure if that's the case for these).

As to naming of the package, I know we currently use `docker-scan-plugin` (for `docker scan`); given that these debs/rpms _do_ install a cli plugin, I'm inclined to name them accordingly; both for consistency, and  because I'm not sure if there's much value in "obscuring" the fact that it is a plugin. I'll leave that choice to @justincormack though.

**Comment by thaJeztah** [comment]
> Depending on whether or not these will be dynamically linked, we need to either build for each distro+distro-version+architecture, or build once, and duplicate the packages for each one.

Hmmm.. so "build once and copy" won't work, as the version (including distro etc) is also part of the package's metadata, so I guess we'll have to build for each distro (or at least _package_ for each distro). Of course; not sure if that's an actual issue if the binary is exactly the same for all of them (i.e., upgrading distro would be fine with keeping the "old" package).

Here's what's in the `control` file in side the package (compared between the buildx package and the docker-ce-cli) - I haven't checked yet for RPMs;

```bash
cat ./bin/docker-buildx_0.7.0~9-g1250d945_amd64/control/control
Package: docker-buildx
Version: 0:0.7.0~9-g1250d945
Section: default
Priority: optional
Architecture: amd64
Maintainer: Docker <support@docker.com>
Vendor: Docker, Inc
Installed-Size: 51408
Homepage: https://github.com/docker/buildx
Description: Docker Buildx plugin extends build capabilities with BuildKit.
```


```bash
cat ./bin/docker-ce-cli_20.10.11~3-0~ubuntu-focal_amd64/control/control
Package: docker-ce-cli
Source: docker-ce
Version: 5:20.10.11~3-0~ubuntu-focal
Architecture: amd64
Maintainer: Docker <support@docker.com>
Installed-Size: 153865
Depends: libc6 (>= 2.4)
Recommends: docker-scan-plugin
Conflicts: docker (<< 1.5~), docker-engine, docker-engine-cs, docker.io, lxc-docker, lxc-docker-virtual-package
Breaks: docker-ce (<< 5:0)
Replaces: docker-ce (<< 5:0)
Section: admin
Priority: optional
Homepage: https://www.docker.com
Description: Docker CLI: the open-source application container engine
 Docker is a product for you to build, ship and run any application as a
 lightweight container
 .
 Docker containers are both hardware-agnostic and platform-agnostic. This means
 they can run anywhere, from your laptop to the largest cloud compute instance and
 everything in between - and they don't require you to use a particular
 language, framework or packaging system. That makes them great building blocks
 for deploying and scaling web apps, databases, and backend services without
 depending on a particular stack or provider.
```

File: hack/nfpm.yml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+platform: linux
+
+version: ${VERSION}
+epoch: 0
+
+section: default
+
+maintainer: Docker <support@docker.com>
+description: Docker Buildx plugin extends build capabilities with BuildKit.
+vendor: ${VENDOR}
```
[the remark this answers is not in the document]
Comment (crazy-max) [reply hack/nfpm.yml:16]: > Do you know if `maintainer` sets `Packager` in the RPM packages?

Yes it is: https://github.com/goreleaser/nfpm/blob/1de91949f55eae477f999e45171586a26b702109/rpm/rpm.go#L231

For deb or apk, maintainer is associated with the right field, [`maintainer` for apk](https://github.com/goreleaser/nfpm/blob/1de91949f55eae477f999e45171586a26b702109/apk/apk.go#L553) and [`Maintainer` for deb](https://github.com/goreleaser/nfpm/blob/1de91949f55eae477f999e45171586a26b702109/deb/deb.go#L676) so I guess that's a mistake on NFPM to use that field as Packager for rpm :(

[the remark this answers is not in the document]
Comment (crazy-max) [reply hack/nfpm.yml:16]: I will open a PR on NFPM repo to switch between vendor/packager.

[the remark this answers is not in the document]
Comment (crazy-max) [reply hack/nfpm.yml:16]: We might be able to add an override on NFPM repo for this case.

[the remark this answers is not in the document]
Comment (crazy-max) [reply hack/nfpm.yml:16]: PR has been merged and `PACKAGER` arg has been added. Here is the result:

```console
$ rpm -qip bin/docker-buildx-0.7.0~9-g1250d945.m.x86_64.rpm 
Name        : docker-buildx
Epoch       : 0
Version     : 0.7.0~9-g1250d945.m
Release     : 1
Architecture: x86_64
Install Date: (not installed)
Group       : Tools/Docker
Size        : 47112192
License     : Apache-2.0
Signature   : (none)
Source RPM  : docker-buildx-0.7.0~9-g1250d945.m-1.src.rpm
Build Date  : Thu Dec 16 11:31:29 2021
Build Host  : buildkitsandbox
Relocations : (not relocatable)
Packager    : Docker <support@docker.com>
Vendor      : Docker
URL         : https://github.com/docker/buildx
Summary     : Docker Buildx plugin extends build capabilities with BuildKit.
Description :
Docker Buildx plugin extends build capabilities with BuildKit.
```

**Comment by crazy-max** [comment]
@thaJeztah 

> So IIUC, the idea is for the buildx repository to provide the "recipe" for building the `.deb` and `.rpm` packages, and to update the `docker-ce-packaging` pipeline, to run it as part of releases, correct?

Yes that's correct

> I had a look at what it produces, and noticed that this currently produces static binaries (packaged as `.deb` / `.rpm`).
> 
> ```shell
> ldd /usr/libexec/docker/cli-plugins/docker-buildx
> 	linux-vdso.so.1 (0x00007fffc1dc7000)
> 	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fa7e9f0a000)
> 	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fa7e9b19000)
> 	/lib64/ld-linux-x86-64.so.2 (0x00007fa7ea129000)
> ```
> 
> I know the current buildx binaries we ship in the `docker-ce-cli` packages are dynamically linked, and perhaps that's not an issue, but I do recall that @justincormack didn't want to be using static binaries (and I also recall some binaries could not be shipped as static binaries on Docker Desktop, related to some GPL code, but not sure if that's the case for these).

Not sure but I think it depends on the situation. For a cli plugin like buildx it doesn't make much sense and in the contrary adds a lot of overhead imo.

> As to naming of the package, I know we currently use `docker-scan-plugin` (for `docker scan`); given that these debs/rpms _do_ install a cli plugin, I'm inclined to name them accordingly; both for consistency, and because I'm not sure if there's much value in "obscuring" the fact that it is a plugin. I'll leave that choice to @justincormack though.

Sure let me know if it needs to be changed: https://github.com/crazy-max/buildx/blob/d24a0cdae3563b04d7f1a4e3cee93efa0605b2c1/hack/nfpm.yml#L1

> * We need to double-check if there's differences in the `.deb` / `.rpm` metadata that may be different between distro versions. Admitted; the install for this plugin is more straightforward than (e.g.) `dockerd` or `containerd`, but I know at some point our release-pipeline tried to build packages using a single distro, and there were incompatibilities between `.deb` packages built for newer distros when installed on older versions of the distro (see

For a cli plugin pkg which only "extracts" without using any triggers or pre(post)install scripts it doesn't seem we can hit this case.

> * or build once, and duplicate the packages for each one.

Don't think it would work as the distro ver would be included in the release.

> * Have a way to generate the right filename for each distro/distro-version

We can follow the Debian version scheme sure. Smth like `docker-buildx_0.7.0~deb11u0` or `docker-buildx_0.7.0-el8` for rpm pkgs.

> Given that `buildx` is currently stashed in the `docker-ce-cli` package, and this will move it to a separate package, we need to add the correct metadata for this, so that they cannot be installed side-by-side (which would overwrite each other's binaries)

Good point

> Hmmm.. so "build once and copy" won't work, as the version (including distro etc) is also part of the package's metadata, so I guess we'll have to build for each distro (or at least _package_ for each distro). Of course; not sure if that's an actual issue if the binary is exactly the same for all of them (i.e., upgrading distro would be fine with keeping the "old" package).

I can add a specific ARG to change the release based on the distro.

> Here's what's in the `control` file in side the package (compared between the buildx package and the docker-ce-cli) - I haven't checked yet for RPMs;

### deb

```console
$ dpkg-deb -I bin/docker-buildx_0.7.0~9-g1250d945.m_amd64.deb 
 new Debian package, version 2.0.
 size 12219872 bytes: control archive=419 bytes.
       1 bytes,     0 lines      conffiles
     306 bytes,    10 lines      control
      81 bytes,     1 lines      md5sums
 Package: docker-buildx
 Version: 0:0.7.0~9-g1250d945.m
 Section: default
 Priority: optional
 Architecture: amd64
 Maintainer: Docker <support@docker.com>
 Vendor: Docker
 Installed-Size: 46008
 Homepage: https://github.com/docker/buildx
 Description: Docker Buildx plugin extends build capabilities with BuildKit.
```

Compared with docker-scan-plugin pkg:

```console
$ dpkg-deb -I bin/docker-scan-plugin_0.12.0~debian-bullseye_amd64.deb 
 new Debian package, version 2.0.
 size 3518216 bytes: control archive=556 bytes.
     340 bytes,    11 lines      control
     164 bytes,     2 lines      md5sums
 Package: docker-scan-plugin
 Source: docker-ce (5:20.10.12~3-0~debian-bullseye)
 Version: 0.12.0~debian-bullseye
 Architecture: amd64
 Maintainer: Docker <support@docker.com>
 Installed-Size: 12946
 Enhances: docker-ce-cli
 Section: admin
 Priority: optional
 Homepage: https://github.com/docker/scan-cli-plugin
 Description: Docker scan cli plugin.
```

## rpm

```console
$ rpm -qip bin/docker-buildx-0.7.0~9-g1250d945.m.x86_64.rpm 
Name        : docker-buildx
Epoch       : 0
Version     : 0.7.0~9-g1250d945.m
Release     : 1
Architecture: x86_64
Install Date: (not installed)
Group       : Tools/Docker
Size        : 47112192
License     : Apache-2.0
Signature   : (none)
Source RPM  : docker-buildx-0.7.0~9-g1250d945.m-1.src.rpm
Build Date  : Thu Dec 16 11:31:29 2021
Build Host  : buildkitsandbox
Relocations : (not relocatable)
Packager    : Docker <support@docker.com>
Vendor      : Docker
URL         : https://github.com/docker/buildx
Summary     : Docker Buildx plugin extends build capabilities with BuildKit.
Description :
Docker Buildx plugin extends build capabilities with BuildKit.
```

Compared with docker-scan-plugin pkg:

```console
$ rpm -qip bin/docker-scan-plugin-0.12.0-3.el8.x86_64.rpm 
warning: bin/docker-scan-plugin-0.12.0-3.el8.x86_64.rpm: Header V4 RSA/SHA512 Signature, key ID 621e9f35: NOKEY
Name        : docker-scan-plugin
Epoch       : 0
Version     : 0.12.0
Release     : 3.el8
Architecture: x86_64
Install Date: (not installed)
Group       : Tools/Docker
Size        : 13247224
License     : ASL 2.0
Signature   : RSA/SHA512, Mon Dec 13 15:21:53 2021, Key ID c52feb6b621e9f35
Source RPM  : docker-scan-plugin-0.12.0-3.el8.src.rpm
Build Date  : Mon Dec 13 12:45:11 2021
Build Host  : c4ca28c21cdf
Relocations : (not relocatable)
Packager    : Docker <support@docker.com>
Vendor      : Docker
URL         : https://github.com/docker/scan-cli-plugin/
Summary     : Docker Scan plugin for the Docker CLI
Description :
Docker Scan plugin for the Docker CLI.
```

**Comment by crazy-max** [comment]
@thaJeztah 

> > Hmmm.. so "build once and copy" won't work, as the version (including distro etc) is also part of the package's metadata, so I guess we'll have to build for each distro (or at least _package_ for each distro). Of course; not sure if that's an actual issue if the binary is exactly the same for all of them (i.e., upgrading distro would be fine with keeping the "old" package).
> 
> I can add a specific ARG to change the release based on the distro.

I added the possibility to define a release version for each distro. See https://github.com/crazy-max/buildx/blob/23d8eaffa41e2748b0b7894bb239290352745c4b/docker-bake.hcl#L10-L15

**Comment by crazy-max** [comment]
@thaJeztah Last commit publish pkgs to `docker/buildx-pkg`. Will look like this: https://hub.docker.com/r/crazymax/buildx-pkg/tags

image

Structure inside the scratch image:

image

**Comment by crazy-max** [comment]
Changed pkg name to `docker-buildx-plugin` as discussed. PTAL @thaJeztah @tonistiigi.

**Pull request closed without merging**